### PR TITLE
CORDA-2157 Hash to  Signature constraints migration

### PR DESCRIPTION
**Pull request by josecoll**
This implementation follows a variation of the two approaches defined in the "[Migration from hash to signature constraint"](https://github.com/corda/corda-private/blob/496a1a366622d7dac8c7a52e3a2d01bdb39e971b/docs/source/design/data-model-upgrades/migrate-to-signature-constraint.md) design document.

Specifically, 
- both the original (used by Hash constrained states) and signed (used by Signature constrained states) Contract JAR must be CZ whitelisted
- the contract package namespace used in the signed Contract JAR must be registered in the Network Parameters.

Constraint transitioning validation checks are then made to ensure:
1. Hash-codes of respective Contract Jars (unsigned/signed) are registered in CZ whitelist
2. Signature (eg. public key) of Signed Jar matches the signature (eg. public key) used to register the Java package namespace.
3. Signature constraint specified keys (in transaction outputs) match the Signed Jar keys.

**I think both the above checks are excessive and require additional operational overhead, so perhaps we should agree to relax one of them. Please comment in this PR.**

The ledger transaction will contain both signed/unsigned attachments (by version) for all associated input/output states.

This PR also introduces the notion of a 'version' identifier associated with Contract Attachment Jars (stored in the database) and associated Contract Jar CorDapps (stored in the jar manifest). 
Further changes are required to populate and use this version id in transaction building and ledger verification (to be completed in https://r3-cev.atlassian.net/browse/CORDA-2150).

**Note**:  An original objective was to implement a function that returned the hash of the signed version of the original hash constrained jar, and verify it matches the original. Experimentation proved this not to be a reproducible task (see summary in https://r3-cev.atlassian.net/browse/CORDA-2156 and code in `colljos-spike-jar-stripping-hashing` branch)

### CODE REVIEW

**Comment by josecoll** [comment]
@chrisr3 can you review the changes I applied to the `core-deterministic` code.
As mentioned on Slack, I'm not clear on why the `core-deterministic` code is still using the Deprecated API in `WireTransaction.toLedgerTransaction`. For now, I've made changes to accommodate continued usage (and had to introduce another Deprecated variant of the same function to maintain API stability), but I think we need to decide on how best to proceed here.
Note, we have:
1) x2 Deprecated methods (used by `core-deterministic`)
2) Recommended method used by everything else but annotated as @DeleteForDJVM

File: core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt [hunk 1]
```diff
@@ -191,8 +193,17 @@ data class LedgerTransaction @JvmOverloads constructor(
             val contractAttachment = contractAttachmentsByContract[state.contract]
                     ?: throw TransactionVerificationException.MissingAttachmentRejection(id, state.contract)
 
-            val constraintAttachment = AttachmentWithContext(contractAttachment, state.contract,
-                    networkParameters?.whitelistedContractImplementations)
+            checkNotNull(networkParameters)
+            val constraintAttachment =
+                if (state.constraint is HashAttachmentConstraint) {
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/transactions/LedgerTransaction.kt]: what is the difference between the 2 branches?

**Comment by tudor-malene** [comment]
Hi @josecoll, is this PR ready for review?

I don't think it actually works with the current code. I don't see how both constraints can pass.

Also, this breaks the no-overlap rule - 2 attachments for the same contract are being passed to the LedgerTransaction.attachments.  I don't see any code that disables the implementation of the no-overlap rule.
After I merge the AttachmentsClassloader work,  some work needs to be done there to select only one of the 2 attachments to verify the transaction.

I don't quite understand either why it needs both CZ whitelisted and package ownership?

**Comment by josecoll** [comment]
@tudor-malene as pointed out in the PR description, we may need both of CZ whitelisted and package ownership, but remember we are not implementing the signed/unsigned hashcode checking function, so package ownership solely may not be sufficient to prove provenance/identity.

**Comment by josecoll** [comment]
`Also, this breaks the no-overlap rule - 2 attachments for the same contract are being passed to the LedgerTransaction.attachments. I don't see any code that disables the implementation of the no-overlap rule.`

Not sure I follow ... different inputs for the same transaction may have different versions of that same Contract (and hence multiple attachments).

**Comment by josecoll** [comment]
>> I don't think it actually works with the current code. I don't see how both constraints can pass.

Not sure I follow ... it will pass if it satisfies the `canBeTransitionedFrom` rules.

**Comment by josecoll** [comment]
I am actually going to add more Test Coverage once we decide what the final transition rules should be (eg. checking CZ hashes and/or package ownership)

**Comment by tudor-malene** [comment]
I would suggest trying a  manual (or behave) test scenario that actually attempts to migrate a state from Hash to Sig constraint on real nodes.

**Comment by tudor-malene** [comment]
> `Also, this breaks the no-overlap rule - 2 attachments for the same contract are being passed to the LedgerTransaction.attachments. I don't see any code that disables the implementation of the no-overlap rule.`
> 
> Not sure I follow ... different inputs for the same transaction may have different versions of that same Contract (and hence multiple attachments).

Currently the no-overlap rule states that a transaction can only have 1 attachment for a type of contract ( it doesn't matter what transactions created the input states).
See: https://docs.corda.net/api-contract-constraints.html

**Comment by tudor-malene** [comment]
> > > I don't think it actually works with the current code. I don't see how both constraints can pass.
> 
> Not sure I follow ... it will pass if it satisfies the `canBeTransitionedFrom` rules.

The transitioning is an additional rule. 
But firstly, every contract attachment needs to satisfy the AttachmentConstraints of all input and output states.

**Comment by tudor-malene** [comment]
> @tudor-malene as pointed out in the PR description, we may not need both of CZ whitelisted and package ownership, but remember we are not implementing the signed/unsigned hashcode checking function, so package ownership solely may not be sufficient to prove provenance/identity.

I realise that :). 
I was just not sure what the whitelisting would prove. 
That the signed and unsigned jar are actually the same jar?
How would you differentiate between normal whitelisting and whitelisting for this purpose?

File: core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt [hunk 2]
```diff
@@ -39,13 +40,15 @@ interface AttachmentConstraint {
      * Rules:
      *  * It is allowed for output states to inherit the exact same constraint as the input states.
      *  * The [AlwaysAcceptAttachmentConstraint] is not allowed to transition to a different constraint, as that could be used to hide malicious behaviour.
-     *  * Nothing can be migrated from the [HashAttachmentConstraint] except a [HashAttachmentConstraint] with the same hash.
      *  * Anything (except the [AlwaysAcceptAttachmentConstraint]) can be transitioned to a [HashAttachmentConstraint].
      *  * You can transition from the [WhitelistedByZoneAttachmentConstraint] to the [SignatureAttachmentConstraint] only if all signers of the JAR are required to sign in the future.
+     *  * You can transition form a [HashAttachmentConstraint] to a [SignatureAttachmentConstraint] when the following conditions are met:
```
Comment (chrisr3) [inline core/src/main/kotlin/net/corda/core/contracts/AttachmentConstraint.kt]: "form" -> "from"

File: .ci/api-current.txt [hunk 3]
```diff
@@ -433,6 +433,7 @@ public interface net.corda.core.contracts.Attachment extends net.corda.core.cont
   @NotNull
   public abstract java.util.List<java.security.PublicKey> getSigners()
   public abstract int getSize()
+  public abstract String getVersion()
```
[the remark this answers is not in the document]
Comment (josecoll) [reply .ci/api-current.txt]: Over zealous - I've confined to adding this to the `ContractAttachment` implementation of `Attachment`.

**Comment by szymonsztuka** [comment]
> > `Also, this breaks the no-overlap rule - 2 attachments for the same contract are being passed to the LedgerTransaction.attachments. I don't see any code that disables the implementation of the no-overlap rule.`
> > Not sure I follow ... different inputs for the same transaction may have different versions of that same Contract (and hence multiple attachments).
> 
> Currently the no-overlap rule states that a transaction can only have 1 attachment for a type of contract ( it doesn't matter what transactions created the input states).
> See: https://docs.corda.net/api-contract-constraints.html
> 
> > > I would question if this is still valid.

I think it's both valid and obsolete:

So far for hash constraints and zone whitelist constraints there the transaction had only one Contract implementation class/JAR (for hash constraint by default the same JAR created input stated and was running the current transaction , as hash need to be the same), for zone whitelist constraints in was external change that from day x now Contract implementation is y (so in the system the old JAR was replaced by new one),  for signature constraint we allow an input generated by older Contact class implementation (JAR) to take part in transaction run by newer Contract Class implementation (JAR),  the no-overlapping rule is still valid - the output states needs to be tun by the same Contract Class Implementation (JAR)  but the inputs could be created by older ones.  Now at lease for non-downgrade rule we need to have "older" JAR (to carry their versions) accessible in our transaction . 
`WireTransaction.toLedgerTransactionInternal (...` passes now more attachments - the extra ones are the JARs which created input states - and they will be duplicated - they are of the same `ContractClassNames` as the Attachment/JAR to run contracts for this transaction. 

The problem is now any code inside TransactionBuilder needs to know the overlapped Attachments are    to check transition rules from inputs states, and for now (thanks to signature constraints) an input state could be created by different "version" of ContractClassName than the one running this transaction.

**Comment by tudor-malene** [comment]
> I think it's both valid and obsolete:
> 
> So far for hash constraints and zone whitelist constraints there the transaction had only one Contract implementation class/JAR (for hash constraint by default the same JAR created input stated and was running the current transaction , as hash need to be the same), for zone whitelist constraints in was external change that from day x now Contract implementation is y (so in the system the old JAR was replaced by new one), 

This is not actually how it works. Transactions need to specify an attachment for each contract type. That attachment needs to satisfy the constraints on all input and output states.
There can be only one because if there are multiple it becomes impossible to build a deterministic environment to verify that transaction in.
The Whitelist constraint did't force any change from day x. It just introduced a newer version of the code. Nothing stops node selecting the old version. 

> for signature constraint we allow an input generated by older Contact class implementation (JAR) to take part in transaction run by newer Contract Class implementation (JAR), 

This is true for the whitelist constraint too.

> the no-overlapping rule is still valid - the output states needs to be tun by the same Contract Class Implementation (JAR) but the inputs could be created by older ones. Now at lease for non-downgrade rule we need to have "older" JAR (to carry their versions) accessible in our transaction .
> `WireTransaction.toLedgerTransactionInternal (...` passes now more attachments - the extra ones are the JARs which created input states - and they will be duplicated - they are of the same `ContractClassNames` as the Attachment/JAR to run contracts for this transaction.

The thing is you don't care about those jars when you verify the transaction. All you need is for those states to be deserializable in the classloader of the transaction.

> 
> The problem is now any code inside TransactionBuilder needs to know the overlapped Attachments are to check transition rules from inputs states, and for now (thanks to signature constraints) an input state could be created by different "version" of ContractClassName than the one running this transaction.

Again, the signature constraint is no different from the Whitelist constraint. Both allow a state to be created and spent with different versions.

**Comment by josecoll** [comment]
Superseded  by new implementation https://github.com/corda/corda/pull/4261

**Pull request closed without merging**